lib/core/src/sync/mod.rs

```diff
@@ -1 +1,329 @@
+        remote_url: String,
+        persister: Arc<Persister>,
+        signer: Arc<Box<dyn Signer>>,
+        client: Box<dyn SyncerClient>,
+    ) -> Self {
+        Self {
+            remote_url,
+            persister,
+            signer,
+            client,
+        }
+    }
+
+    fn commit_record(
+        &self,
+        decrypted_record: &DecryptedRecord,
+        sync_state: SyncState,
+        is_update: bool,
+        last_commit_time: Option<u32>,
+    ) -> Result<()> {
+        match decrypted_record.data.clone() {
+            SyncData::Chain(chain_data) => self.persister.commit_incoming_chain_swap(
+                &chain_data,
+                sync_state,
+                is_update,
+                last_commit_time,
+            )?,
+            SyncData::Send(send_data) => self.persister.commit_incoming_send_swap(
+                &send_data,
+                sync_state,
+                is_update,
+                last_commit_time,
+            )?,
+            SyncData::Receive(receive_data) => self.persister.commit_incoming_receive_swap(
+                &receive_data,
+                sync_state,
+                is_update,
+                last_commit_time,
+            )?,
+        }
+        Ok(())
+    }
+
+    fn load_sync_data(&self, data_id: &str, record_type: RecordType) -> Result<SyncData> {
+        let data = match record_type {
+            RecordType::Receive => {
+                let receive_data: ReceiveSyncData = self
+                    .persister
+                    .fetch_receive_swap_by_id(data_id)?
+                    .ok_or(anyhow!("Could not find Receive swap {data_id}"))?
+                    .into();
+                SyncData::Receive(receive_data)
+            }
+            RecordType::Send => {
+                let send_data: SendSyncData = self
+                    .persister
+                    .fetch_send_swap_by_id(data_id)?
+                    .ok_or(anyhow!("Could not find Send swap {data_id}"))?
+                    .into();
+                SyncData::Send(send_data)
+            }
+            RecordType::Chain => {
+                let chain_data: ChainSyncData = self
+                    .persister
+                    .fetch_chain_swap_by_id(data_id)?
+                    .ok_or(anyhow!("Could not find Chain swap {data_id}"))?
+                    .into();
+                SyncData::Chain(chain_data)
+            }
+        };
+        Ok(data)
+    }
+
+    async fn fetch_and_save_records(&self) -> Result<()> {
+        let latest_revision = self
+            .persister
+            .get_sync_settings()?
+            .latest_revision
+            .unwrap_or(0);
+        let req = ListChangesRequest::new(latest_revision, self.signer.clone())?;
+        let incoming_records = self.client.pull(req).await?.changes;
+
+        self.persister.set_incoming_records(&incoming_records)?;
+        let latest_revision = incoming_records.last().map(|record| record.revision);
```

Comment:
Just to check, the last record is always the highest revision?

Reply:
Yes; at least that's what we expect from the server